### PR TITLE
[DataGrid] Add `getFilterState` method

### LINKED ISSUE #13413: [data grid] multiple data sets state management & filtering
### The problem in depth

Hi everyone!

I would like to share some struggles I've been having with MUI X Data Grid Pro recently and, hopefully, someone much more experienced than I could point me in the right direction or, unfortunately, just confirm my not so optimistic suspicions about the current capabilities of the Data Grid.

So, I have a functionality which is something like this:
A user logs into their account and sees a dashboard which has 5 Data Grid components below (only 1 shown at a time).

Each of those components opens in a separate tab (my app is a SPA, so they are toggled via ToggleGroup, also by MUI).
Those tabs show different Data Grid components which are based on `assetType` (which are 'webService', 'domain', 'ssl', 'ip' and 'port'), but anyway, essentially, we have just 5 different components (each using Data Grid), shown separately (one at a time).

Each table also has 5 tabs of their own, 4 of which are based on the asset status filter.
Initial data is without a status filter and is shown on the `all` tab. The rest are corresponding to statuses `new`, `inProgress`, `fixed` and `accepted`.

The most interesting part comes in inside of each of those components, however.

This is an example of my custom `SslTable`:
```
//imports
interface IInfrastructureScanningDomainTableProperties {
  domain: IInfrastructurePageDomain;
  tableData: IInfrastructurePageSslTableView[];
}

export const InfrastructurePageSslTable: FunctionComponent<IInfrastructureScanningDomainTableProperties> = observer(
  ({ domain, tableData }) => {
    tableData = tableData ?? [];

    // make data copy to separate rows views & counters filtered by statuses logic
    const rows = [...tableData];

    const apiRef = useGridApiRef();
    const theme = useTheme();

    const [filterButtonEl, setFilterButtonEl] = React.useState<HTMLButtonElement | null>(null);
    const [selectedRows, setSelectedRows] = React.useState<GridRowId[]>([]);

    const isHaveResults = tableData.length !== 0;

    // filter data copy and NOT the original data
    const filteredData = useMemo(() => {
      const memoizedFilteredData = rows.filter(
      (entity) => entity.verificationStatus === domain.ui.sslTableView.entity.filter || domain.ui.sslTableView.entity.filter === 'all')
      return memoizedFilteredData;
    }, [rows, domain.ui.sslTableView.entity.filter, domain.ui.sslTableView.entity.textFilter]);
    
    const columns = useMemo(() => [
      //...rest of the columns,
      InfrastructurePageCommonVerificationStatusColumn(domain, '80px'),
    ], []);

    const filterColumns = ({ field, columns, currentFilters }: FilterColumnsArgs) => {
      const filteredFields = currentFilters?.map((item) => item.field);
      return columns
        .filter((colDef) => colDef.filterable && (colDef.field === field || !filteredFields.includes(colDef.field)))
        .map((column) => column.field);
    };

    const getColumnForNewFilter = ({ currentFilters, columns }: GetColumnForNewFilterArgs) => {
      if (currentFilters.length >= 5) {
        return null;
      }
      const filteredFields = currentFilters?.map(({ field }) => field);
      const columnForNewFilter = columns
        .filter((colDef) => colDef.filterable && !filteredFields.includes(colDef.field))
        .find((colDef) => colDef.filterOperators?.length);
      return columnForNewFilter?.field ?? null;
    };

    const handlePaginationModelChange = () => {
      const dashboardElement = document.getElementById("sticky-area-container");
      if (dashboardElement) {
        setTimeout(() => dashboardElement.scrollIntoView({ block: 'end' }), 100)
      }
    };

    // DO NOT CHANGE params & dependencies
    const debouncedHandleStateChange = useCallback(debounce(() => {
      domain.ui.handleStateChange(rows, 'ssl', apiRef);
    }, 500), [apiRef]);

    return (
      <Grid xs={12} lg={12}>
        <Box>
          <Box sx={{ minWidth: 0, maxWidth: '700px', position: 'relative' }}>
            <InfrastructurePageFilterToggleGroup assetType={'ssl'} data={tableData} domain={domain} />
            {
              !!selectedRows.length &&
                <Box>
                  <CustomMultiselect selectedRows={selectedRows} domain={domain} />
                </Box>
            }
          </Box>
          <Box sx={{ display: 'flex', minWidth: 0, width: '323px', gap: '8px', marginLeft: 'auto' }}/>
        </Box>
        <Grid mt={'-12px'} sx={{ minWidth: 0, width: '1328px', height: '60vh' }}>
          {isHaveResults && (
            <DataGridPro
              apiRef={apiRef}
              // USE onFilterModelChange and NOT onStateChange!!!
              onFilterModelChange={debouncedHandleStateChange}
              localeText={localeText}
              checkboxSelection
              rows={filteredData}
              columns={columns}
              pagination
              onPaginationModelChange={handlePaginationModelChange}
              onRowSelectionModelChange={setSelectedRows}
              initialState={initialState}
              //rest of Data Grid Pro props
            />
          )}
        </Grid>
        <Grid mt={'10vh'}></Grid>
      </Grid>
    );
  },
);
```

And this is `domain.ui.handleStateChange` function logic which is called inside the `debouncedHandleStateChange` function:
```
handleStateChange (tableData: any, assetType: string, apiRef: React.MutableRefObject<GridApiPro>): Function {
    let filteredDataRows = tableData;
    
    if (apiRef.current?.state?.filter?.filteredRowsLookup) {
      const dataGridRowIDs = new Set(
        Object.entries(apiRef.current.state.filter.filteredRowsLookup)
          .filter(([_, value]) => value === true)
          .map(([key, _]) => key)
      );
      
      // DO NOT REMOVE!!!
      // asset data & filter counters change for Data Grid onFilterModelChange
      
      const filteredDataRowsByAll = [...tableData.filter((item) => dataGridRowIDs.has(item.id))].filter((item) => this[`${assetType}TableView`].entity.filter === 'all');
      this[`${assetType}TableViewRowsCountByAll`].setValue(filteredDataRowsByAll.length);
      
      const filteredDataRowsByNew = [...tableData.filter((item) => dataGridRowIDs.has(item.id))].filter((item) => item.verificationStatus === ENTITY_VERIFICATION_STATUS.new);
      this[`${assetType}TableViewRowsCountByNew`].setValue(filteredDataRowsByNew.length);
      
      const filteredDataRowsByInProgress = [...tableData.filter((item) => dataGridRowIDs.has(item.id))].filter((item) => item.verificationStatus === ENTITY_VERIFICATION_STATUS.inProgress);
      this[`${assetType}TableViewRowsCountByInProgress`].setValue(filteredDataRowsByInProgress.length);
      
      const filteredDataRowsByFixed = [...tableData.filter((item) => dataGridRowIDs.has(item.id))].filter((item) => item.verificationStatus === ENTITY_VERIFICATION_STATUS.fixed);
      this[`${assetType}TableViewRowsCountByFixed`].setValue(filteredDataRowsByFixed.length);
      
      const filteredDataRowsByAccepted = [...tableData.filter((item) => dataGridRowIDs.has(item.id))].filter((item) => item.verificationStatus === ENTITY_VERIFICATION_STATUS.accepted);
      this[`${assetType}TableViewRowsCountByAccepted`].setValue(filteredDataRowsByAccepted.length);
    }
    
    return filteredDataRows;
}
```

This is a custom `ToggleGroup` component which I'm using to display each of the filtered data tabs with corresponding counters:
```
//imports
interface IInfrastructurePageFilterToggleGroup {
  domain: IInfrastructurePageDomain;
  data:
    | IInfrastructurePageDomainTableView[]
    | IInfrastructurePageIpAddressesTableView[]
    | IInfrastructurePagePortTableView[]
    | IInfrastructurePageSslTableView[];
  assetType: keyof IInfrastructureTableViewFiltersByAssetCategory;
}

export const InfrastructurePageFilterToggleGroup: FunctionComponent<IInfrastructurePageFilterToggleGroup> = observer(
  ({ domain, data, assetType }) => {
    const value = domain.ui[`${assetType}TableView`]?.entity.filter || 'all';

    const filterData = (data: any[], filterBy: InfrastructureAssetVerificationStatuses | 'all') =>
      data.filter(
        (
          entity:
            | IInfrastructurePageWebServiceTableView
            | IInfrastructurePageDomainTableView
            | IInfrastructurePageIpAddressesTableView
            | IInfrastructurePagePortTableView
            | IInfrastructurePageSslTableView
            | IInfrastructurePageResponsiblePersonWithGroupingTableView,
        ) => entity.verificationStatus === filterBy,
      ).length || 0;

    const ignoredCount = filterData(data, ENTITY_VERIFICATION_STATUS.ignore);

    const All = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='All'
        filterBy='all'
        isActive={'all' === value}
        numberOfEntity={domain.ui[`${assetType}TableViewRowsCountByAll`].value}
      />
    )
    
    const New = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='New'
        filterBy={ENTITY_VERIFICATION_STATUS.new}
        isActive={ENTITY_VERIFICATION_STATUS.new === value}
        numberOfEntity={domain.ui[`${assetType}TableViewRowsCountByNew`].value}
      />
    )
    
    const inProgress = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='In Progress'
        filterBy={ENTITY_VERIFICATION_STATUS.inProgress}
        isActive={ENTITY_VERIFICATION_STATUS.inProgress === value}
        numberOfEntity={domain.ui[`${assetType}TableViewRowsCountByInProgress`].value}
      />
    )
    
    const Accepted = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='Accepted'
        filterBy={ENTITY_VERIFICATION_STATUS.accepted}
        isActive={ENTITY_VERIFICATION_STATUS.accepted === value}
        numberOfEntity={domain.ui[`${assetType}TableViewRowsCountByAccepted`].value}
      />
    )
    
    const Ignore = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='Ignored'
        filterBy={ENTITY_VERIFICATION_STATUS.ignore}
        isActive={ENTITY_VERIFICATION_STATUS.ignore === value}
        numberOfEntity={ignoredCount}
      />
    )
    
    const Fixed = (
      <CreateInfrastructurePageFilterToggleGroupItem 
        name='Fixed'
        filterBy={ENTITY_VERIFICATION_STATUS.fixed}
        isActive={ENTITY_VERIFICATION_STATUS.fixed === value}
        numberOfEntity={domain.ui[`${assetType}TableViewRowsCountByFixed`].value}
      />
    )

    return (
      <Grid container sx={{ height: '60px', width: '615px' }}>
          <ToggleButtonGroup
            value={value}
            exclusive
            onChange={(event) => domain.ui.filterByStatusHandler(assetType, event)}
            aria-label="text alignment"
          >
            {All}
            {New}
            {inProgress}
            {Fixed}
            {Accepted}
            {Ignore}
          </ToggleButtonGroup>
      </Grid>
    );
  },
);

interface IInfrastructurePageFilterToggleGroupItemProps {
  name: string,
  filterBy: InfrastructureAssetVerificationStatuses | 'all',
  isActive: boolean,
  numberOfEntity: number,
}

const CreateInfrastructurePageFilterToggleGroupItem = observer(({
  name,
  filterBy,
  isActive,
  numberOfEntity,
}: IInfrastructurePageFilterToggleGroupItemProps) => {
  const theme: Theme = useTheme();
  return (
    <ToggleButton      
      aria-selected={isActive}
      value={filterBy}
    >
      <Box
      >
        <Typography variant="body2" color={isActive ? 'text.primary' : theme.palette.text.secondary}>
          {name}
        </Typography>
      </Box>
      <EasmBadge numberOfEntity={numberOfEntity} />
    </ToggleButton>
  );
});
```

And this is the `filterByStatusHandler` function which I'm calling while toggling between tabs:
  ```
filterByStatusHandler = (
    assetType: keyof IInfrastructureTableViewFiltersByAssetCategory,
    event: React.MouseEvent<HTMLInputElement>,
  ): void => {
    const {
      currentTarget: { value: newValue },
    } = event;
    if (!newValue) return;
    const entity = this[`${assetType}TableView`]?.entity;
    
    //@ts-ignore
    this[`${assetType}TableView`]?.setEntity({ ...entity, filter: newValue });
    
    return;
};
```

Now, I have a problem — while I'm on the `all` tab (while ```domain.ui[`${assetType}TableView`]?.entity.filter === 'all'```), 
everything is working fine: each tab data is correctly filtered, counters are correctly counted and when I toggle between tabs (only if I start from 'all' tab), data and counters remain correct. 
But if I apply any search or filter in the Data Grid component while on any other tab ('new', 'inProgress', 'fixed' or 'accepted'), 
all tabs data arrays are incorrectly filtered and counters are incorrectly counted (including also breaking filters and counter for 'all' tab data).

I've tried to separately filter data and toggle view for different statuses, but, eventually, either the counters are wrong while updated simultaneously (on any tab other than on 'all'), or, the counters are fine, but only when I pass unfiltered data into the `rows` prop of the Data Grid (but in that case, the views are wrong, since they are not filtered by status).
I've tried this to fix the latter like this:
```
if (this[`${assetType}TableView`].entity.filter === 'all') {
  apiRef.current.updateRows(
    rowIds.map((rowId) => ({ id: rowId, ...filteredDataRowsByAll.find((item) => item.id === rowId) })),
  );
}

if (this[`${assetType}TableView`].entity.filter === ENTITY_VERIFICATION_STATUS.new) {
  apiRef.current.updateRows(
    rowIds.map((rowId) => ({ id: rowId, ...filteredDataRowsByNew.find((item) => item.id === rowId) })),
  );
}

if (this[`${assetType}TableView`].entity.filter === ENTITY_VERIFICATION_STATUS.inProgress) {
  apiRef.current.updateRows(
    rowIds.map((rowId) => ({ id: rowId, ...filteredDataRowsByInProgress.find((item) => item.id === rowId) })),
  );
}

if (this[`${assetType}TableView`].entity.filter === ENTITY_VERIFICATION_STATUS.fixed) {
  apiRef.current.updateRows(
    rowIds.map((rowId) => ({ id: rowId, ...filteredDataRowsByFixed.find((item) => item.id === rowId) })),
  );
}

if (this[`${assetType}TableView`].entity.filter === ENTITY_VERIFICATION_STATUS.accepted) {
  apiRef.current.updateRows(
    rowIds.map((rowId) => ({ id: rowId, ...filteredDataRowsByAccepted.find((item) => item.id === rowId) })),
  );
}
```

So, basically, I just tried to update the rows of the Data Grid for each of the tab manually without affecting the data passed into the `rows` prop of the Data Grid,
which, in turn, affects calculating the counters.

And for that, also, I've tried saving the state from the `api.current.state` while I have unfiltered data on the `all` tab, and count filters for other tabs based on that saved state, and not the `apiRef` changed in the future, but that also didn't work:

```
useEffect(() => {
  const filterValue = domain.ui.sslTableView.entity.filter;
  const newFilterModel: GridFilterModel = {
    items: filterValue === 'all' ? [] : [{ field: 'verificationStatus', operator: 'equals', value: filterValue }],
  };
  setFilterModel(newFilterModel);
}, [domain.ui.sslTableView.entity.filterActive]);

const debouncedHandleStateChange = useCallback(
  debounce(() => {
    if (domain.ui.sslTableView.entity.filter === 'all') {
      const currentState = apiRef.current.state;
      setSavedGridState(currentState);
      domain.ui.handleStateChange(rows, 'ssl', currentState);
    } else {
      if (savedGridState) {
        setTimeout(() => {
          domain.ui.handleStateChange(rows, 'ssl', savedGridState);
        }, 100);
      }
    }
  }, 500),
  [rows, savedGridState, domain.ui.sslTableView.entity.filterActive]
);
```

So, my guess is — either what I'm trying to achieve is not achievable based on how MUI X Data Grid Pro works right now,

or, if I'm wrong and/or maybe am close to a working solution — I would very much appreciate any help and advice.

Thanks!

p.s. 

I'm already using MobX for states in my logic domain stores (all those `this.ui.tableView.entity` and ```this[`${assetType}TableView`]?.setEntity({ ...entity, filter: newValue })``` lines with states and methods).

But I also have been thinking of trying to use `React Context API` for this case, just couldn't wrap my head around of how to do that properly in this case and how it would actually help with the Data Grid underlying issue, of when any change in the data passed into the `rows` prop of the Data Grid component is causing changes of filter counters on all tabs.

I also tried `zustand`, but it hasn't changed much with this issue in particular, unfortunately.

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```
I used Brave browser (it's Chromium based)

System:
    OS: macOS 14.4.1
  Binaries:
    Node: 18.18.2 - ~/.nvm/versions/node/v18.18.2/bin/node
    npm: 9.8.1 - ~/src/frontend/node_modules/.bin/npm
    pnpm: 9.2.0 - ~/.nvm/versions/node/v16.18.0/bin/pnpm
  Browsers:
    Chrome: 125.0.6422.142
    Edge: Not Found
    Safari: 17.4.1
    Brave: 1.66.118 Chromium: 125.0.6422.147 (Official Build) (arm64) 
  npmPackages:
    @emotion/react: 11.11.3 => 11.11.3
    @emotion/styled: 11.11.0 => 11.11.0
    @mui/base: 5.0.0-beta.36 => 5.0.0-beta.36
    @mui/core-downloads-tracker:  5.15.10
    @mui/icons-material: 5.15.10 => 5.15.10
    @mui/lab: 5.0.0-alpha.132 => 5.0.0-alpha.132
    @mui/material: 5.15.10 => 5.15.10
    @mui/private-theming:  5.15.9
    @mui/styled-engine:  5.15.9
    @mui/styled-engine-sc: 6.0.0-alpha.16 => 6.0.0-alpha.16
    @mui/styles:  5.14.12
    @mui/system: 5.15.9 => 5.15.9
    @mui/types: 7.2.13 => 7.2.13
    @mui/utils:  5.15.9
    @mui/x-data-grid: 6.18.1 => 6.18.1
    @mui/x-data-grid-generator: 6.18.1 => 6.18.1
    @mui/x-data-grid-premium:  6.18.1
    @mui/x-data-grid-pro: 6.18.1 => 6.18.1
    @mui/x-date-pickers: 6.5.0 => 6.5.0
    @mui/x-license-pro:  6.10.2
    @types/react: 18.2.23 => 18.2.23
    react: 18.2.0 => 18.2.0
    react-dom: 18.2.0 => 18.2.0
    styled-components: 6.0.0-rc.1 => 6.0.0-rc.1
    typescript: 4.3.5 => 4.3.5
```
</details>

**Search keywords**: state management filtering conditional data for rows
**Order ID**: 89674

### PR DESCRIPTION
**Pull request by cherniavskii**
Demo: https://deploy-preview-13418--material-ui-x.netlify.app/x/react-data-grid/filtering-recipes/#calculating-row-count-in-advance

Fixes #13413

### CODE REVIEW

File: docs/data/data-grid/filtering-recipes/filtering-recipes.md [hunk 1]
```diff
@@ -13,3 +13,7 @@ Currently if you want to use the [Quick filter](/x/react-data-grid/filtering/qui
 A common use case is to have certain components positioned outside of the grid. Because of the way the grid context works this might not be a straightforward thing to do. The example below illustrates how this use case can be achieved.
 
 {{"demo": "QuickFilterOutsideOfGrid.js", "bg": "inline", "defaultCodeOpen": false}}
+
+## Calculating row count in advance
+
```
Comment (michelengelen) [inline docs/data/data-grid/filtering-recipes/filtering-recipes.md:18]: Should we maybe add some description here?

File: docs/data/data-grid/filtering-recipes/FilteredRowCount.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+export default function FilteredRowCount() {
+  const { data } = useDemoData({
+    dataSet: 'Commodity',
+    rowLength: 1000,
+    maxColumns: 10,
+  });
+
+  const apiRef = useGridApiRef();
+
+  const [predefinedFiltersRowCount, setPredefinedFiltersRowCount] = React.useState<
```
Comment (michelengelen) [inline docs/data/data-grid/filtering-recipes/FilteredRowCount.tsx:42]: I would probably prefer an object here, but I guess this is mainly personal preference. As long as the function is pure it should be fine.

**Review by michelengelen — APPROVED** [review]
Just a few nits! Looks pretty solid! :shipit:

File: docs/data/data-grid/filtering-recipes/filtering-recipes.md [hunk 3]
```diff
@@ -13,3 +13,7 @@ Currently if you want to use the [Quick filter](/x/react-data-grid/filtering/qui
...
+{{"demo": "FilteredRowCount.js", "bg": "inline", "defaultCodeOpen": false}}
```
Comment (michelengelen) [inline docs/data/data-grid/filtering-recipes/filtering-recipes.md:19]: Hey @cherniavskii 
I am no allowed to push to your branch, so I'll just leave a suggestion here that you can apply

```suggestion

## Calculating row count in advance

In many data exploration scenarios, it's highly beneficial to display the row count associated with a predefined filter upfront.
This transparency empowers users to make informed decisions about which filter to apply, potentially saving them time and effort.

The [Grid API](/x/react-data-grid/api-object/#how-to-use-the-api-object) provides the [`getFilterState`](/x/api/data-grid/grid-api/#grid-api-prop-getFilterState) function, which allows you to retrieve the filter state for a given filter model without applying it to the data grid. By effectively utilizing `getFilterState`, you can dynamically calculate the row count that would result from applying a specific predefined filter.

{{"demo": "FilteredRowCount.js", "bg": "inline", "defaultCodeOpen": false}}
```

**Review by michelengelen — CHANGES_REQUESTED** [review]
sry for the inconvenience! 🙇🏼

File: packages/x-data-grid/src/models/api/gridFilterApi.ts [hunk 4]
```diff
@@ -61,4 +62,10 @@ export interface GridFilterApi {
    * Returns the value of the `ignoreDiacritics` prop.
    */
   ignoreDiacritics: DataGridProcessedProps['ignoreDiacritics'];
+  /**
+   * Returns the filter state for the given filter model without applying it to the data grid.
+   * @param {GridFilterModel} filterModel The filter model to get the state for.
+   * @returns {GridStateCommunity['filter']} The filter state.
+   */
+  getFilterState: (filterModel: GridFilterModel) => GridStateCommunity['filter'];
```
Comment (romgrk) [inline packages/x-data-grid/src/models/api/gridFilterApi.ts:70]: This sounds like it should be a (pure) selector, but I see how that would be annoying to implement.

**Review by romgrk — APPROVED** [review]

File: docs/data/data-grid/filtering-recipes/filtering-recipes.md (hunk 3 above)
Comment (michelengelen) [reply docs/data/data-grid/filtering-recipes/filtering-recipes.md]: @cherniavskii gentle ping 🔔

**Comment by cherniavskii** [comment]
@michelengelen I need an approval from you in order to merge it

**Pull request merged**